### DISCUSSION TITLE
Model for uninterpreted functions

### DISCUSSION THREAD

**Discussion by kfriedberger** [cvc5/cvc5#9086] [Q&A] [answered]
Is there a programmatic way to get a model evaluation for an uninterpreted function?

For example, lets use some symbols `var:Int`, `boundVar:Int`, an uninterpreted function `foo:Int->Int` 
and the assertion `var = 1 & EXISTS boundVar : (boundVar = 2 & var = foo(boundVar))`.

I want to receive a model like `{var:=1; foo(2):=1}` from the solver.
The symbol `boundVar` can be ignored, because it is quantified and not in the scope for the complete assertion.

We can analyse the assertion for symbols and then query the solver for evaluations for each symbol.
This seems to work only for simple Int/Bool/... symbols (i.e., we receive `var:=1`), but not for uninterpreted functions.
Is there a way to query models for  uninterpreted functions?

Our first target is quantifier-free logic, the next step is a usage of the quantifier operation EXISTS.
The situation might be more complex when using a quantifier operation FORALL.

**Comment by kfriedberger** [comment]
Second example:
For a more assertion like `var = 1 & foo(var) = 4 & foo(12) = 7 & foo(34) = 9` it would be nice to have a model like `{var:=1; foo(1):=4; foo(12):=7; foo(34):=9}`. Something like `{var:=1; foo(_x):=ITE(_x=1, 4, ITE(_x=12, 7, ITE(_x=34, 9, 0)))}` with a user-given const symbol `_x:Int` would also be possible, i.e., the user could query a model evaluation for `foo(_x)` and get a ITE-cascade (ITE:IfThenElse).

I noticed, that CVC5 already provides such an ITE-cascade via the toString method of the UF's first child term. Is it possible to access that ITE-cascade programmatically?

**Reply by mpreiner** [reply]
@kfriedberger Did you try the [getValue() API method](https://cvc5.github.io/docs/cvc5-1.0.2/api/java/io/github/cvc5/Solver.html#getValue(io.github.cvc5.Term))? This will return the ITE structure for UFs.

**Comment by kfriedberger** [comment]
@mpreiner
I already thought of this method, as It would be the ideal place for such a functionality.
But could not figure out, how to use it. Can you give more details about using this method for UFs?
How does the term need to look like? Do I need to build a new term (Kind.UF_APPLY, MY_UF, NEW_PARAM) with fresh constant symbol for the parameter?

**Comment by kfriedberger** [comment, marked as answer]
Ok, I though too complex. Simply using the UF as-is is sufficient. Thanks for your hint. I am one step further now.

**Reply by mpreiner** [reply]
Excatly, you only have to pass the UF itself to getValue().

**Comment by kfriedberger** [comment]
One more question:
The lambda-term does not list all possible/used parameter values, but ignores one of them and uses it as a 'default' parameter in the last else-case. Is there a way to get a listing of all used parameter values for an UF?

Example:
The assertion `foo(1)=2 & foo(3)=4` results in a UF-evaluation `foo := arg1 -> ITE(arg1=1; 2; 4)` which ignores the parameter value `3`. This is sound, but the user does not see that `foo(3)` is important (a user might want to apply further steps where this information is important).

Would it be possible to enrich the UF-model with all used assignments, e.g. , like `foo := arg1 -> ITE(arg1=1; 2; (ITE(arg2=3; 4; X))` where `X` can be an arbitrary value, for example `4` if that is simple to support.

**Reply by kfriedberger** [reply]
@mpreiner could you have a look at the latest question here?

**Comment by ajreynol** [comment]
At the moment, it isn't possible to distinguish the necessary values of function inputs as you suggest.

Indeed, having a placeholder `X` might be highly useful to distinguish which values are not important. 

FYI, we do have a coarse grained feature `--model-cores=MODE` which drops the definitions of variables in models whose values are not important based on some criteria.  Doing what you suggest would require a refinement where *parts* of definitions are dropped, which would also be interesting.  I will keep you posted if we support this in the future.
